app/services/scholarsphere_location_operator.rb

```diff
@@ -0,0 +1,77 @@
+# frozen_string_literal: true
+
+class ScholarsphereLocationOperator
+  def initialize(params:, token:)
+    @params = params
+    @token = token
+  end
+
+  def perform
+    if !@token.write_access?
+      operator_response(401, 'unauthorized')
+    elsif !params_valid?
+      operator_response(422, 'params_invalid')
+    elsif update_open_access_location.present?
+      operator_response(200, 'success')
+    elsif @existing_location
+      operator_response(422, 'existing_location')
+    else
+      operator_response(404, 'no_content')
+    end
+  end
+
+  private
+
+    attr_reader :params, :token
+
+    def params_valid?
+      return false if params[:scholarsphere_open_access_url].blank?
+
+      !(params[:activity_insight_id].present? && params[:doi].present?)
+    end
+
+    def update_open_access_location
+      locations = []
+
+      ActiveRecord::Base.transaction do
+        filtered_publications.each do |publication|
+          return if existing_location?(publication)
+
+          location = find_or_create_location(publication)
+          locations << location.url if location.valid?
+        end
+      end
+
+      locations
+    end
+
+    def filtered_publications
+      publications = []
+
+      if params[:activity_insight_id].present?
+        publications = Publication.filter_by_activity_insight_id(Publication, params[:activity_insight_id])
+      elsif params[:doi].present?
+        publications = Publication.filter_by_doi(Publication, params[:doi])
+      end
+
+      publications
+    end
+
+    def existing_location?(publication)
+      @existing_location ||= publication.open_access_locations
+        .filter { |l| l.source == Source::USER }
+        .index_by(&:url)
+        .key?(params[:scholarsphere_open_access_url])
+    end
+
+    def find_or_create_location(publication)
+      publication.open_access_locations.find_or_create_by(
+        source: Source::USER,
```

Comment:
Here and in the method above I think we should be using `Source::SCHOLARSPHERE` instead of `source::USER`. `Source::USER` is for `OpenAccessLocation`s that are created by faculty/researcher users submitting a URL via the web form. It does feel a little odd to create an `OpenAccessLocation` with `Source::SCHOLARSPHERE` here since it could be any API user submitting the request and not necessarily the ScholarSphere app (or someone acting on its behalf). However, the param that's being submitted is `scholarsphere_open_access_url`, and in the old data model we would have been pushing the value into the field of the same name on the `Publication` which was also used by the ScholarSphere importer and the ScholarSphere upload workflow in RMD. In the new data model, instead of a value in that field, we'd have an associated `OpenAccessLocation` with a `source` of `Source::SCHOLARSPHERE` and a value in its `url` field.

So we'll probably be overloading the meaning of `Source::SCHOLARSPHERE` a little bit, but I'm not sure that it really matters. We could introduce a new `source` value like `Source::API_USER` or something like that, but it would complicate this even further and probably wouldn't be necessary.